k2/python/k2/fsa_algo.py

```diff
@@ -145,3 +164,100 @@ def arc_sort(fsa: Fsa) -> Fsa:
+        a_fsa=a_fsa.arcs,
+        b_fsa=b_fsa.arcs,
+        a_aux_labels=a_fsa.aux_labels.to(torch.int32),
+        b_aux_labels=b_fsa.aux_labels.to(torch.int32),
+        need_arc_map=need_arc_map)
+
+    # Some of entries in a_arc_map and b_arc_map may be zero.
+    # We add 1 here so that every entry is non-negative.
+    a_arc_map = a_arc_map.to(torch.int64) + 1
+    b_arc_map = b_arc_map.to(torch.int64) + 1
+
+    out_fsa = Fsa.from_ragged_arc(ragged_arc)
+    for name, a_value in a_fsa.named_tensor_attr():
+        if name == 'aux_labels':
+            # aux_label is handled specially
+            continue
+        if hasattr(b_fsa, name):
+            b_value = getattr(b_fsa, name)
+
+            # a_arc_map and b_arc_map have been offset by 1
+            # so we need a padding here
+            padding = a_value.new_zeros((1, *a_value.shape[1:]))
+            a_value = torch.cat((padding, a_value), dim=0)
+            b_value = torch.cat((padding, b_value), dim=0)
+
+            assert a_value.dtype == b_value.dtype == torch.float32
+            value = a_value.index_select(0, a_arc_map) \
+                    + b_value.index_select(0, b_arc_map)
+            setattr(out_fsa, name, value)
+
+    if hasattr(b_fsa, 'aux_labels'):
+        b_aux_labels = b_fsa.aux_labels
+        padding = b_aux_labels.new_zeros(1)
+        b_aux_labels = torch.cat((padding, b_aux_labels), dim=0)
+        out_fsa.aux_labels = b_aux_labels.index_select(0, b_arc_map)
+
+    for name, a_value in a_fsa.named_non_tensor_attr():
+        if name == 'osym':
+            continue
+        setattr(out_fsa, name, a_value)
+
+    for name, b_value in b_fsa.named_non_tensor_attr():
+        if not hasattr(out_fsa, name):
+            setattr(out_fsa, name, b_value)
+
+    return out_fsa
+
+
+def shortest_distance(fsa: Fsa) -> torch.Tensor:
+    '''Return the shortest distance from the start state to the final state
+    in the tropical semiring.
+
+    Note:
+      It uses the opposite sign. That is, It uses `max` instead of `min`.
+
+    Args:
+      fsa:
+        The intpu FSA.
+    Returns:
+      The sum of scores along the best path.
+    '''
+    need_arc_map = True
+    _, arc_map = _k2.shortest_distance(fsa.arcs, need_arc_map)
+    arc_map = arc_map.to(torch.int64)
+
+    return fsa.score[arc_map].sum()
+
+
+def shortest_path(fsa: Fsa) -> Fsa:
+    need_arc_map = True
+    ragged_arc, arc_map = _k2.shortest_path(fsa.arcs, need_arc_map)
+    arc_map = arc_map.to(torch.int64)  # required by index_select
+    out_fsa = Fsa.from_ragged_arc(ragged_arc)
+    for name, value in fsa.named_tensor_attr():
+        setattr(out_fsa, name, value.index_select(0, arc_map))
+    for name, value in fsa.named_non_tensor_attr():
+        setattr(out_fsa, name, deepcopy(value))
+    return out_fsa
+
+
+def project_output(fsa: Fsa) -> Fsa:
+    out_fsa = Fsa.from_ragged_arc(fsa.arcs)
+    indexes = out_fsa.arcs.values()[:, 2] != -1
```

Comment:
can you please comment here, IDK what this is doing.

Reply:
... if you are trying to exclude final-states here; is that necessary?  I thought we already had -1's in the aux_labels.

Reply:
For an arc:
```
src_state  dest_state label score
```

This function replaces `label` with `aux_label`, so the arc becomes

```
src_state dest_state aux_label score
```

But for arcs entering the final state, their labels are `-1` and `aux_label` 0. we do not want to change these arcs.

Reply:
So it would be easier to set the aux_label for those arcs to -1.

Reply:
.. then there will be no mismatch.  we can add code to check that -1's were not added or removed.

Reply:
> I thought we already had -1's in the aux_labels.

Yes, we are using `-1` currently because I use `compose` instead of `intersect`.
I remember that you suggested the aux_labels for arcs entering the final state should
be 0, not -1. I'll will change the aux_labels to 0 once compose is replaced by intersect.

Reply:
... anyway, I think this can just be done as an in-place modification of the FSA, doesn't have to be an operation.  We can have GetSymbols() and SetSymbols(), that would respectively get and set the symbols on the arcs; and we could just do `fsa.SetSymbols(fsa['aux']);`

Reply:
I wasn't sure about that decision.  Let's set them to -1 for those final-arcs, now that it seems to be a problem to have them as zero.

Reply:
Now arcs entering the final state have -1 for its labels and aux_labels.

project_output is removed.